File: Kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

class GildedRose(var items: Array<Item>) {

    fun updateQuality() = items.forEach { item ->
        item.takeIf { it.name != "Sulfuras, Hand of Ragnaros" }?.apply { sellIn -= 1 }

        when {
            item.name.startsWith("Conjured") -> item.decreaseQualityBy(2)
            item.sellIn < 0 && item.name == "Backstage passes to a TAFKAL80ETC concert" -> item.quality = 0
            item.name == "Backstage passes to a TAFKAL80ETC concert" && item.sellIn < 6 -> item.increaseQualityBy(3)
            item.name == "Backstage passes to a TAFKAL80ETC concert" && item.sellIn < 11 -> item.increaseQualityBy(2)
            item.name == "Backstage passes to a TAFKAL80ETC concert" -> item.increaseQualityBy(1)
            item.name == "Aged Brie" -> item.increaseQualityBy(1)
            item.name == "Sulfuras, Hand of Ragnaros" -> item.quality = item.quality
            item.sellIn < 0 -> item.decreaseQualityBy(2)
            else -> item.decreaseQualityBy(1)
        }
    }
}

fun Item.decreaseQualityBy(decreaseFactor: Int) {
    this.quality.takeIf { decreaseFactor > 0 && (this.quality - decreaseFactor) >= 0 }
            ?.also {
                this.quality -= decreaseFactor
            }
}

fun Item.increaseQualityBy(increaseFactor: Int) {
    this.quality.takeIf { increaseFactor > 0 && (this.quality + increaseFactor) <= 50 }
            ?.also {
                this.quality += increaseFactor
            }
}

File: clojure/src/gildedrose_refactoring_kata/gilded_rose.clj
(ns gildedrose-refactoring-kata.gilded_rose
  (:require [gildedrose-refactoring-kata.item :refer :all]))

(defn update_quality
  "update quality in list of items"
  [items]
  (map #((comp
          (fn [item] (if (and (not (= (:name item) "Aged Brie")) (not (= (:name item) "Backstage passes to a TAFKAL80ETC concert")))
                       (if (> (:quality item) 0)
                         (if (not (= (:name item) "Sulfuras, Hand of Ragnaros")) (update item :quality dec) (item))
                         item)
                       (if (< (:quality item) 50)
                         ((comp
                           (fn [x] (update x :quality inc))
                           (fn [x] (if (= (:name x) "Backstage passes to a TAFKAL80ETC concert")
                                     ((comp
                                       (fn [y] (if (< (:sell_in y) 11)
                                                 (if (< (:quality y) 50) (update y :quality inc) y)
                                                 y))
                                       (fn [y] (if (< (:sell_in y) 6)
                                                 (if (< (:quality y) 50) (update y :quality inc) y)
                                                 y)))
                                      x)
                                     x)))
                          item)
                         item)))
          (fn [item] (if (not (= (:name item) "Sulfuras, Hand of Ragnaros")) (update item :sell_in dec) item))
          (fn [item] (if (< (:sell_in item) 0)
                       (if (not (= (:name item) "Aged Brie"))
                         (if (and (not (= (:name item) "Aged Brie")) (not (= (:name item) "Backstage passes to a TAFKAL80ETC concert")))
                           (if (> (:quality item) 0)
                             (if (not (= (:name item) "Sulfuras, Hand of Ragnaros"))
                               (update item :quality dec)
                               item)
                             item)
                           (if (< (:quality item) 50)
                             (update item :quality inc)
                             item))
                         item)
                       item)))
         %)
       items))